### PR TITLE
Fixes #32224 - Expose Katello events for Webhooks

### PR DESCRIPTION
**Pull request by ofedoren**
This PR exposes new events for subscriprions within Foreman Webhooks plugin, to test this you might want to have it installed.

@lzap, you might want to test it and probably tell what is missing.

TODO:
 - [x] Add DSL docs
 
 NOTE: I've changed the order of initializers because currently the actions are being loaded after all plugin registration methods are called, which leads to addition of new observable events being impossible. Also `extend_observable_events` is available from `2.4`, so I've updated the required version as well.

### CODE REVIEW

**Comment by lzap** [comment]
@jjeffers what are our chances getting this into 4.0? cheers :-)

**Comment by jjeffers** [comment]
> @jjeffers what are our chances getting this into 4.0? cheers :-)
> 
> I will be testing this tomorrow.

Probably low chance, since we just built rc3.1. The next release should be the GA.

**Comment by lzap** [comment]
Yeah that's what I thought. I was also hoping the patch would be even smaller, it looks like an initializer change is needed. Okay.

File: lib/katello/engine.rb
Comment (lzap) [inline lib/katello/engine.rb:105]: I am not sure I understand the consequences of this, but since I'd like to ask for minor release inclusion for this patch I wonder if @adamruzicka would be able to confirm this is safe to do.

File: app/lib/actions/katello/actions_jail.rb
Comment (lzap) [inline app/lib/actions/katello/actions_jail.rb:33]: I am wondering how a concern will work in this case, class Jail is defined via the concern and then redefined later. I guess I need to test it.

By the way is the reason you inherit `::Actions::ObservableAction::Jail` instead of `Safemode::Jail` to also have `task` allowed?

**Comment by lzap** [comment]
Nicely done, can you add those DSL docs? I am starting with testing right away. I was hoping that we only add a single line into plugin.rb that initializer reordering is unfortunate but I can test some edge-case scenarios (e.g. dynflow resume?).

File: lib/katello/engine.rb
Comment (adamruzicka) [reply lib/katello/engine.rb:105]: This looks reasonable to me and I don't see any immediate footguns. @jlsherrill anything I'm overlooking?

**Comment by ofedoren** [comment]
@lzap, working on the docs, I've updated the PR so it's testable with Safemode enabled (I forgot that when I was writing this I had that disabled)

File: app/lib/actions/katello/actions_jail.rb
Comment (ofedoren) [reply app/lib/actions/katello/actions_jail.rb:40]: Well, I guess for actions we want have `task` object available to get some info. The first action that has that is `host_run_job_succeeded`, so I thought it might be useful for all the actions, thus I inherit `::Actions::ObservableAction::Jail`.

Also, I've updated the way the `Jail` gets possibility to be re-opened after the module is included.

File: app/lib/actions/katello/actions_jail.rb
Comment (lzap) [inline app/lib/actions/katello/actions_jail.rb:37]: I am just wondering if it is not cleaner just to define attributes in this module and simply repeating all allowed attributes for each class. But I am okay with this.

Comment (ofedoren) [reply app/lib/actions/katello/actions_jail.rb:37]: Sure it would be cleaner than some magic Ruby-woodoo, but repeating is not so good as well since one might forget to update allowed methods somewhere else.

**Comment by lzap** [comment]
Rubocop failures.

**Comment by ofedoren** [comment]
@lzap, updated. Test failures don't seem to be related..

File: app/lib/actions/katello/actions_common_jail.rb
Comment (lzap) [inline app/lib/actions/katello/actions_common_jail.rb:62]: This really calls for breaking up into two concerns: `ActionsContentViewJailConcern` and `ActionsOrganizationJailConcern`.

**Review by lzap — CHANGES_REQUESTED** [review]
Other than that, I am good.

**Comment by ofedoren** [comment]
@lzap, updated. I've renamed those a bit to have a structure.

Just a reminder for reviewers: if you want to test docs locally as well:

`FOREMAN_APIPIE_LANGS=en bundle exec rake apipie_dsl:cache` within `foreman` directory.

**Review by lzap — APPROVED** [review]
Thank you I am good.

To Katello reviewers: This patch introduces Jail support for Safemode, in order to expose methods they must be explicitly listed. We tried to find minimum amount of methods which gives reasonable power to webhooks users. We have landed on this rather flat list of methods because the dynflow actions carries the data in a hash.

File: app/lib/actions/katello/jail_concern/content_view.rb
Comment (lzap) [inline app/lib/actions/katello/jail_concern/content_view.rb:41]: I think I found a bug tho, I am getting safemode error for Content View create event, I think this concern is not included in the ContentView model at all. It is not a Dynflow action but ActiveRecord event as well, so it should be just a simple Jail.

```
2021-04-06T12:24:34 [I|app|06a8fc9e] ForemanWebhooks::EventSubscriber: content_view_created.event.foreman event received
2021-04-06T12:24:34 [E|app|06a8fc9e] Safemode::NoMethodError: undefined method '#content_view_id' for Katello::ContentView::Jail (Katello::ContentView)
```

File: app/models/katello/content_view.rb
Comment (lzap) [inline app/models/katello/content_view.rb:94]: I think organization and location (name,label,id) is worth exposing too for CV.

File: app/lib/actions/katello/jail_concern/content_view.rb
Comment (ofedoren) [reply app/lib/actions/katello/jail_concern/content_view.rb:41]: Well, for content view object it should be just `id`, not the `content_view_id`. There is (or at least should be, I don't remember) documentation for Content View class in `all`/`additional` section in DSL docs.

Comment (lzap) [reply app/lib/actions/katello/jail_concern/content_view.rb:41]: So it is used in other Dynflow action(s) but in Content View (CRUD) the Jail is missing, so actually here it is correct.

File: app/models/katello/content_view.rb
Comment (lzap) [reply app/models/katello/content_view.rb:94]: We are missing Jail for this model here.

**Review by lzap — CHANGES_REQUESTED** [review]
Actually one missing jail.

**Comment by ofedoren** [comment]
@lzap, updated. There was Jail for content views already, I've just updated it with `id` and `organization`.

**Comment by lzap** [comment]
> @lzap, updated. There was Jail for content views already, I've just updated it with `id` and `organization`.

Oh my bad I started testing with `id` and I assumed there was no Jail class. Name would be maybe enough, but this is good, let me retest.

File: app/lib/actions/katello/content_view/promote.rb
Comment (lzap) [inline app/lib/actions/katello/content_view/promote.rb:8]: Just for the record, this does not work atm, organization jail is overwritten in the include block by content view.

**Review by lzap — CHANGES_REQUESTED** [review]
We are close, it does not work yet.

File: app/lib/actions/katello/content_view/promote.rb
Comment (lzap) [reply app/lib/actions/katello/content_view/promote.rb:8]: Since the Jail.allow method concatenates to a list, you can call it multiple times. So I suggest you assume Jail already exists, move these includes down after Jail class definition and append from the included block.

**Comment by ofedoren** [comment]
@lzap, thanks for noticing that `const_set` will replace the `Jail` with the second `include`. I've played a little bit more with class reopening and noticed that what works with plain Ruby in `irb` doesn't work in Rails. Well, even if it worked I agree that this complicates simple things and might break something in the future.

I've returned the simple way for `Jail`s and left mixins only for their main purpose.

File: app/lib/actions/katello/content_view/promote.rb
Comment (ofedoren) [reply app/lib/actions/katello/content_view/promote.rb]: I've tried that one, but when I try something like `base::Jail` or `base.const_get(:Jail)` it returns an object that doesn't understand `.inspect` even.

**Review by lzap — APPROVED** [review]
Everything works as expected. Can we have somebody looking on this PR, there are many comments already so folks might be skipping it. @ehelms @jlsherrill thanks.

File: lib/katello/engine.rb
Comment (ehelms) [inline lib/katello/engine.rb:102]: I am confused by the moving around of this code. It appears to not change anything other than where it appears in the file?

File: app/lib/actions/katello/repository/sync.rb
Comment (ehelms) [inline app/lib/actions/katello/repository/sync.rb:161]: Is this tied to `def sync_result` ? Otherwise, a blank line above this would help.

File: app/lib/actions/katello/repository/sync.rb
Comment (ehelms) [inline app/lib/actions/katello/repository/sync.rb:184]: Do I understand right that the `apipie` definition, the include JailConcern and the class Jail declaration are all tied to the same concept? Events?

**Comment by ehelms** [comment]
> To Katello reviewers: This patch introduces Jail support for Safemode, in order to expose methods they must be explicitly listed. We tried to find minimum amount of methods which gives reasonable power to webhooks users. We have landed on this rather flat list of methods because the dynflow actions carries the data in a hash.

Is the addition of Jail support for Safemode directly tied to Webhooks support? Is that something that is a pre-requisite? Does Jail support without "events" make sense for Katello? One thing I am trying to understand from this while looking over the code is as a developer what parts of the changes are what I need to add to produce events? (and which parts are not 'directly' required)

File: lib/katello/engine.rb
Comment (ofedoren) [reply lib/katello/engine.rb:102]: Yes. From that perspective there are no changes at all. But the order does affects if the needed constants are available in plugin registration block.

File: app/lib/actions/katello/repository/sync.rb
Comment (ofedoren) [reply app/lib/actions/katello/repository/sync.rb:162]: Noticed, I've added missing blank lines.

File: app/lib/actions/katello/repository/sync.rb
Comment (ofedoren) [reply app/lib/actions/katello/repository/sync.rb:185]: Yes, for `actions` these are needed to fully support events in Foreman Webhooks that are being fired up if e.g. repo sync has happened.

`apipie` block is for documentation; JailConcern with public methods to use in webhook templates; Jails are to make those methods available in safe mode.

**Comment by ofedoren** [comment]
@ehelms, short answer would be:

If a developer wants to expose events to be subscribed after then simple `include Foreman::ObservableModel` for models or `include ::Actions::ObservableAction` for actions should be enough. Other additions are related, but not required. Although without them there is no much of usage from the user end perspective since everything will be internal.

File: app/lib/actions/katello/repository/sync.rb
Comment (ehelms) [reply app/lib/actions/katello/repository/sync.rb:185]: Thanks. It seems like it would be useful to group all of the pieces related to webhooks together somehow but I'll let Katello core maintainers weigh in on the overall code structure. I appreciate the helpful details understanding. I was worried that it would be hard over time to track what is required to ensure Webhooks work (and subsequently do not break in the future) and what aspects are unrelated. Also, how to help developers who want to add additinal webhook "hooks" into the code know what to add.

**Comment by lzap** [comment]
I can confirm what @ofedoren said, making events observable via foreman_webhooks is really that simple and we have documented this in the devel docs: https://github.com/theforeman/foreman/blob/develop/developer_docs/how_to_create_a_plugin.asciidoc#adding-subscribers

However if you turn on safemode, then the data coming in the webhook templates will be useless because there will be no way of getting any data out of it. Therefore the Jail and apidocs portions of the patch.

**Comment by jlsherrill** [comment]
Wanted to update, i'm in the middle of testing this.  Overall looks good, will update again soon.

**Review by jlsherrill — APPROVED** [review]
ACK, tested the events and they all worked, rendering the template i created in each case.  I'm okay with the overall structure of the code here

**Comment by lzap** [comment]
Thanks, I asked for 4.0.1 in the issue.

**Comment by lzap** [comment]
I see Katello 4.0.1 being pushed to rubygems. I did set target release in RM to at least have a conversation about possibility of this patch to be cherry picked. Now I've sent my presentation recording to Foreman Birthday party where I announce Foreman Webhooks with Katello 4.0 support :-(

https://projects.theforeman.org/versions/1369

Should I move the target release to 4.0.2 with hope it would make the cut? :-)

**Comment by ofedoren** [comment]
@lzap isn't it already there? https://github.com/Katello/katello/commits/KATELLO-4.0 / https://github.com/Katello/katello/compare/4.0.0...4.0.1

**Comment by jjeffers** [comment]
> @lzap isn't it already there? https://github.com/Katello/katello/commits/KATELLO-4.0 / [4.0.0...4.0.1](https://github.com/Katello/katello/compare/4.0.0...4.0.1)

It should be in there - please see https://github.com/Katello/katello/pull/9359

**Comment by lzap** [comment]
Holy smokes, I think I've used all of them. THANKS! :-)

**Pull request merged**
